### PR TITLE
Uniq 

### PR DESCRIPTION
**Pull request by wildcat26**

### CODE REVIEW

File: src/Streamly/Coreutils/Uniq.hs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+
+-- convert stream of char to stream of Array Char by breaking on newline
+
+splitOnNewLine :: (MonadIO m, IsStream t, Monad m) => t m Char -> t m (A.Array Char)
+splitOnNewLine strm = S.splitOnSuffix (== '\n') A.write strm
+
+
+isSpace :: Char -> Bool
+isSpace c = c == ' '
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:77]: `Data.Char` already exports `isSpace`

File: src/Streamly/Coreutils/Uniq.hs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                                             then s ++ [c]
+                                             else [c]
+                                  Nothing -> s ++ [c]
+
+
+
+skpField :: (IsStream t, Monad m) => Int -> A.Array Char -> t m Char
+skpField n arr = S.concatMap (\s -> S.fromList s)
+                 $ S.scan (FL.mkPureId mergeStringChar "")
+                 $ S.fromList
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:99]: Use `toStream`?

Comment (adithyaov) [reply src/Streamly/Coreutils/Uniq.hs:99]: https://composewell.github.io/streamly/master/Streamly-Internal-Memory-Array.html#v:toStream

File: src/Streamly/Coreutils/Uniq.hs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+isSpace :: Char -> Bool
+isSpace c = c == ' '
+
+safeLast :: [a] -> Maybe a
+safeLast [] = Nothing
+safeLast (x:[]) = Just x
+safeLast (x:xl) = safeLast xl
+
+
+mergeStringChar :: String -> Char -> String
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:85]: This is rather inefficient. What are you trying to do here? Maybe there is a better way.

File: src/Streamly/Coreutils/Uniq.hs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                                             then s ++ [c]
+                                             else if isSpace l == False && isSpace c == False
+                                             then s ++ [c]
+                                             else [c]
+                                  Nothing -> s ++ [c]
+
+
+
+skpField :: (IsStream t, Monad m) => Int -> A.Array Char -> t m Char
+skpField n arr = S.concatMap (\s -> S.fromList s)
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:97]: Is this incomplete?

Comment (adithyaov) [reply src/Streamly/Coreutils/Uniq.hs:97]: Also, what do you expect the output of `skpField _ (['a', 'b', 'c', 'd'] :: Array Char)` to be?

Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:97]: Yes, this was incomplete.

File: src/Streamly/Coreutils/Uniq.hs (hunk 2 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:99]: I had misunderstood `S.scan`. I'm using `S.groupsBy` now.

File: src/Streamly/Coreutils/Uniq.hs (hunk 1 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:85]: Oh thanks. I'll use it.

File: src/Streamly/Coreutils/Uniq.hs (hunk 3 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:103]: I want to fold a `t m Char` into `t m String`. Basically, want to do this
`[' ', ' ', 'a', 'b', 'c' ,' ', ' ', 'e']  -> ["  ", "abc", "  ", "e"]` using streams

Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:103]: Then drop a few strings from the stream generated (based on options provided) and then convert the `String` stream back to a `Char` stream.

File: Benchmark/Coreutils.hs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+       , IOMode(..)
+       , openFile)
+
+import Gauge (defaultMain)
+import Gauge.Benchmark (bench, bgroup, nfAppIO, nfIO)
+import Control.Monad.Catch (MonadThrow, MonadCatch)
+
+
+srcFP :: FilePath
+srcFP = "/home/shruti/sorted.txt"
```
Comment (adithyaov) [inline Benchmark/Coreutils.hs:29]: You might want to remove your name?

File: Main.hs [hunk 6]
```diff
@@ -0,0 +1,4 @@
+module Main where
```
Comment (adithyaov) [inline Main.hs:1]: Why is this required?

File: src/Streamly/Coreutils/Uniq.hs (hunk 3 above)
Comment (adithyaov) [reply src/Streamly/Coreutils/Uniq.hs:102]:
```
groupCS :: (Monad m, IsStream t) => t m Char -> t m String
groupCS = S.groupsBy xorCS FL.toList
  where
    xorCS a b = (isSpace a && isSpace b) || (not (isSpace a || isSpace b))
```
Use this instead.

File: src/Streamly/Coreutils/Uniq.hs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+mergeStringChar s c = case safeHead s of
+                                  Just l -> if eqCh l c == True
+                                            then s ++ [c]
+                                            else [c]
+                                  Nothing -> s ++ [c]
+
+
+
+
+skpField :: (IsStream t, Monad m) => Int -> A.Array Char -> t m Char
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:112]: This is very inefficient. You have all sorts of combinators available to you. Try to make use of them. Try not to use `toList` and `fromList`. Unless you're working explicitly with lists, you won't need to use them.

Comment (adithyaov) [reply src/Streamly/Coreutils/Uniq.hs:112]: What does `skpField` do? I think you can write it in a much simpler way.

Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:112]: > What does `skpField` do? I think you can write it in a much simpler way.

It removes first `skpField opt` number of non-empty strings

File: Main.hs (hunk 6 above)
Comment (wildcat26) [reply Main.hs:1]: Sorry, will remove it.

File: src/Streamly/Coreutils/Uniq.hs (hunk 3 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:102]: I need multiple spaces. :+1:

File: src/Streamly/Coreutils/Uniq.hs (hunk 7 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:112]: > This is very inefficient. You have all sorts of combinators available to you. Try to make use of them. Try not to use `toList` and `fromList`. Unless you're working explicitly with lists, you won't need to use them.

:+1: Used `toStream` to convert an Array to a Stream.
Using `S.concatMap`, I want to convert `t m String` to `t m Char`. Hence, I wanted to use an `S.fromList`.

Is it fine?

Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:112]:
```
{-# INLINE eqCh #-}

eqCh :: Char -> Char -> Bool
eqCh s c = (isSpace s == True && isSpace c == True) || (isSpace s == False && isSpace c == False)



skpField :: (IsStream t, Monad m) => Int -> A.Array Char -> t m Char
skpField n arr = S.concatMap S.fromList
               $ S.drop (2*n)
               $ S.groupsBy eqCh FL.toList
               $ S.cons ' '
               $ AI.toStream arr

-- dropping 2*n works only if there is at least one space in the beginning of a line


```

File: Benchmark/Coreutils.hs (hunk 5 above)
Comment (wildcat26) [reply Benchmark/Coreutils.hs:29]: Should we not push this file at all?

Comment (wildcat26) [reply Benchmark/Coreutils.hs:29]: I will write property based tests and those can be pushed.

File: src/Streamly/Coreutils/Uniq.hs (hunk 7 above)
Comment (adithyaov) [reply src/Streamly/Coreutils/Uniq.hs:94]: Can you give me an example of what it should do? The input and the output.

File: Benchmark/Coreutils.hs (hunk 5 above)
Comment (adithyaov) [reply Benchmark/Coreutils.hs:29]: You should probably use some function to get the current directory.

Comment (wildcat26) [reply Benchmark/Coreutils.hs:29]: Okay. :+1: good idea

File: src/Streamly/Coreutils/Uniq.hs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+defaultUniqOptions :: UniqOptions
+defaultUniqOptions = UniqOptions True False False 0 0 intMax False True '\n'
+
+
+-------------------------------------------------------------------------------
+-- helper functions for uniq
+-------------------------------------------------------------------------------
+
+ignCase :: (IsStream t, Monad m) => Bool -> t m Char -> t m Char
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:52]: This changes the stream itself. We just have to ignore case for the sake of comparision.

Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:52]: Oh, my bad. Thanks :+1:

File: src/Streamly/Coreutils/Uniq.hs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+
+splitOnNewLine :: (MonadIO m, IsStream t, Monad m) => t m Char -> t m String
+splitOnNewLine strm = S.splitOnSuffix (== '\n') FL.toList strm
+
+
+-- Skip X terms
+-- Skip Y chars
+-- Take Z chars
+getRepetition :: (IsStream t, Monad m) => Bool -> Int -> Int -> Int -> t m String -> t m (Int, String)
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:106]: This could take a generic compare function.

File: src/Streamly/Coreutils/Uniq.hs [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+                                                         else (i + 1, a)) (0, ""))
+
+
+onlyDuplicate :: (IsStream t, Monad m) => t m (Int, String) -> t m String
+onlyDuplicate = S.concatMap (\(i, x) -> if i > 1
+                                        then S.replicate i x
+                                        else S.nil)
+
+
+onlyRepeated :: (IsStream t, Monad m) => t m (Int, String) -> t m String
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:119]: Why is the input `t m (Int, String)`? It should be `t m String`?

File: src/Streamly/Coreutils/Uniq.hs [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+onlyRepeated = S.concatMap (\(i, x) -> if i > 1
+                                       then S.replicate 1 x
+                                       else S.nil)
+
+
+onlyUnique :: (IsStream t, Monad m) => t m (Int, String) -> t m String
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:125]: Same as above.

File: src/Streamly/Coreutils/Uniq.hs [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+onlyUnique = S.concatMap (\(i, x) -> if i == 1
+                                     then S.replicate i x
+                                     else S.nil)
+
+
+uniqStrm :: (IsStream t, Monad m, MonadIO m) => Bool -> Int -> Int -> Int -> t m Char -> t m String
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:131]: This could take a generic compare function.

File: src/Streamly/Coreutils/Uniq.hs [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+uniqStrm ign x y z strm = onlyUnique
+   $ getRepetition ign x y z
+   $ splitOnNewLine strm
+
+
+duplicateStrm :: (IsStream t, Monad m, MonadIO m) => Bool -> Int -> Int -> Int -> t m Char -> t m String
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:137]: This could take a generic compare function.

File: src/Streamly/Coreutils/Uniq.hs [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+duplicateStrm ign x y z strm = onlyDuplicate
+   $ getRepetition ign x y z
+   $ splitOnNewLine strm
+
+
+repeatedStrm :: (IsStream t, Monad m, MonadIO m) => Bool -> Int -> Int -> Int -> t m Char -> t m String
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:143]: This could take a generic compare function.

File: src/Streamly/Coreutils/Uniq.hs (hunk 10 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:119]: `uniqStrm`, `duplicateStrm`, `repeatedStrm` have signatures like `t m Char -> t m String`

If the input here would have been `t m String`, then I would also have to pass the `Bool -> Int -> Int -> Int` arguments for `getRepetition`.

Comment (adithyaov) [reply src/Streamly/Coreutils/Uniq.hs:120]:
```
onlyRepeated :: (String -> String -> Bool) -> t m String -> t m (Int, String)
onlyRepeated comparator str = 
    S.concatMap (\(i, x) -> 
        if i > 1
        then S.replicate 1 x
        else S.nil)
    . getRpetition
```
?
Similarly, with `onlyDuplicate`.
This function is too specific to be a top-level function.
You can either change it like so or put it on the `where` clause of `uniqStrm`
Please do the same for `onlyUnique` and `onlyDuplicate` as well.

File: src/Streamly/Coreutils/Uniq.hs [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+repeatedStrm ign x y z strm = onlyRepeated
+   $ getRepetition ign x y z
+   $ splitOnNewLine strm
+
+
+displayCounts :: (IsStream t, Monad m, MonadIO m) => Bool -> Int -> Int -> Int -> ((Int, String) -> Bool) -> t m Char -> t m (Int, String)
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs]: What is this used for?

File: src/Streamly/Coreutils/Uniq.hs [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+compareWith :: (String -> String -> Bool) -> String -> String -> Bool
+compareWith eq s1 s2 = eq s1 s2
+
+
+-- Skip X terms
+-- Skip Y chars
+-- Take Z chars
+compareXYZ :: Bool -> Int -> Int -> Int -> String -> String -> Bool
+compareXYZ ignCase x y z str1 str2 =
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:88]: This can be under `compareUsingOptions`' `where` clause?

File: src/Streamly/Coreutils/Uniq.hs [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+        then compareIgnCase
+        else (==)
+
+
+
+compareUsingOptions :: UniqOptions -> (String -> String -> Bool)
+compareUsingOptions opt = compareXYZ (ignoreCase opt) (skipFields opt) (skipChar opt) (checkChar opt)
+
+
+-- Skip X terms
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:104]: Remove this comment?

File: src/Streamly/Coreutils/Uniq.hs [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+duplicateStrm comparator strm = onlyDuplicate
+   $ getRepetition comparator strm
+
+
+repeatedStrm :: (IsStream t, Monad m, MonadIO m) => (String -> String -> Bool) -> t m String -> t m String
+repeatedStrm comparator strm = onlyRepeated
+   $ getRepetition comparator strm
+
+
+displayCounts :: (IsStream t, Monad m, MonadIO m) => (String -> String -> Bool) -> ((Int, String) -> Bool) -> t m String -> t m (Int, String)
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:147]: What is this used for?

File: src/Streamly/Coreutils/Uniq.hs (hunk 17 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:104]: Yes. :+1:

File: src/Streamly/Coreutils/Uniq.hs (hunk 18 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:147]: When we want to provide counts to the user, at that time, the returned stream should have type `t m (Int, String)`

Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:147]: & the predicate `((Int, String) -> Bool)` is to filter which kind of lines (unique, repeated, distinct) is needed in the output stream

File: src/Streamly/Coreutils/Uniq.hs (hunk 10 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:120]: okay. :+1:

File: src/Streamly/Coreutils/Uniq.hs [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+                     (drop (ix2 + y) str2)
+
+
+-- |
+-- Applies the comparison function to adjacent strings in the stream
+-- and returns a @(Int, String)@ stream where the integer
+-- is the count of occurrences of the string
+--
+-- @since 0.1.0.0
+getRepetition :: (IsStream t, Monad m) => (String -> String -> Bool) -> t m String -> t m (Int, String)
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:175]: Use `UniqResult` here?

File: src/Streamly/Coreutils/Uniq.hs [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+
+-- |
+-- Data type of uniq's output.
+-- @Val s@ represents the occurence of string @s@
+-- in the output stream.
+-- @Count n s@ represents that the string @s@ occurs @n@
+-- times in the output stream.
+--
+-- @since 0.1.0.0
+data UniqResult = Val String
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:37]: Get rid of `Val`?

File: src/Streamly/Coreutils/Uniq.hs [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+    let
+        val (_, x) = Val x
+        eq n (i, _) = i == n
+        ge n (i, _) = i > n
+        cnt (n, s) = Count n s
+        convert predicate = S.map cnt
+            $ S.filter predicate
+            $ getRepetition (compareUsingOptions opt) strm
+    in
+        if count opt == True
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:200]: You don't need `== True`

Comment (adithyaov) [reply src/Streamly/Coreutils/Uniq.hs:200]: What is this if statement for? You can remove it. You can remove the option itself.

File: src/Streamly/Coreutils/Uniq.hs [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+import Streamly.Internal.Data.Stream.StreamK.Type (IsStream)
+
+
+-- |
+-- Data type to capture the output of the stream -
+-- the stream should either be composed of
+-- unique, repeated or duplicate strings
+--
+-- @since 0.1.0.0
+data Output = Unique
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:24]: You can maybe add `All` to this that would just be `getRepetition`

File: src/Streamly/Coreutils/Uniq.hs [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+
+
+
+-- |
+-- Record to specify command line options like count,
+-- skip-fields, check-chars used in GNU uniq
+--
+-- @since 0.1.0.0
+data UniqOptions = UniqOptions {
+                       count :: Bool
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:57]: Remove this.

File: src/Streamly/Coreutils/Uniq.hs [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+                _ -> convert (ge 1)
+        else
+            case (output opt) of
+                Unique -> S.map val
+                     $ S.filter (eq 1)
+                     $ getRepetition (compareUsingOptions opt) strm
+                Repeated -> S.map val
+                     $ S.filter (ge 1)
+                     $ getRepetition (compareUsingOptions opt) strm
+                Duplicate -> S.concatMap (\(i, x) -> S.replicate i (Val x))
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:213]: You don't need to use `concatMap` here. Instead add a function that converts `t m UniqResult` to `t m String`.

File: src/Streamly/Coreutils/Uniq.hs (hunk 19 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:175]: okay.

File: src/Streamly/Coreutils/Uniq.hs (hunk 20 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:37]: Why?

File: src/Streamly/Coreutils/Uniq.hs (hunk 21 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:200]: If `count` is `True`, then the `UniqResult` stream returned will have `Count Int String` elements.
Otherwise, it'll have `Val String` elements.

File: src/Streamly/Coreutils/Uniq.hs (hunk 22 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:24]: okay. :+1: good idea

File: src/Streamly/Coreutils/Uniq.hs (hunk 23 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:57]: why?

File: src/Streamly/Coreutils/Uniq.hs (hunk 24 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:213]: why? Aren't we leaving the final result to `t m UniqResult` ?

File: src/Streamly/Coreutils/Uniq.hs [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+-- @since 0.1.0.0
+data UniqResult = Count Int String
+
+
+-- |
+-- Show instance for @UniqResult@
+--
+-- @since 0.1.0.0
+instance Show UniqResult where
+    show (Count n s) = show $ show n ++ " " ++ s ++ "\n"
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:46]: `show $` is not needed.

File: src/Streamly/Coreutils/Uniq.hs [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+    let
+        eq n (Count i _) = i == n
+        ge n (Count i _) = i > n
+        extract predicate = S.filter predicate
+            $ getRepetition (compareUsingOptions opt) strm
+    in
+        case (output opt) of
+            Unique -> extract (eq 1)
+            All -> extract (\x -> True)
+            _ -> extract (ge 1)
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:195]: What's the difference between `Repeated` and `Duplicate`?

File: Benchmark/Coreutils.hs [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+import Streamly.Internal.Data.Stream.Serial (SerialT(..))
+import Streamly.Internal.Data.Stream.StreamK.Type (IsStream)
+
+srcFP ::IO FilePath
+srcFP = do
+            home <- getHomeDirectory
+            return $ home ++ "/sorted.txt"
+
+dstFP :: IO FilePath
+dstFP = do
```
Comment (adithyaov) [inline Benchmark/Coreutils.hs:32]: https://github.com/composewell/haskell-dev/blob/master/coding-style.rst#do-blocks

File: Benchmark/Coreutils.hs [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+
+import Gauge (defaultMain)
+import System.Directory (getHomeDirectory)
+import Control.Monad.Catch (MonadThrow, MonadCatch)
+import Gauge.Benchmark (bench, bgroup, nfAppIO, nfIO)
+import Streamly.Internal.Data.Stream.Serial (SerialT(..))
+import Streamly.Internal.Data.Stream.StreamK.Type (IsStream)
+
+srcFP ::IO FilePath
+srcFP = do
```
Comment (adithyaov) [inline Benchmark/Coreutils.hs:27]: https://github.com/composewell/haskell-dev/blob/master/coding-style.rst#do-blocks

File: Benchmark/Coreutils.hs [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+charStrm :: (IsStream t, Monad m, MonadIO m, MonadCatch m) => FilePath -> t m Char
+charStrm fp = Un.decodeLatin1 $ File.toBytes fp
+
+
+splitOnNewLine :: (MonadIO m, IsStream t, Monad m) => t m Char -> t m String
+splitOnNewLine strm = S.splitOnSuffix (== '\n') FL.toList strm
+
+
+opt :: UniqOptions
+opt = defaultUniqOptions { skipFields = 12
```
Comment (adithyaov) [inline Benchmark/Coreutils.hs:47]: https://github.com/composewell/haskell-dev/blob/master/coding-style.rst#data-declarations

File: Benchmark/Coreutils.hs [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+                         , skipChar = 10 }
+
+main :: IO ()
+main = do
```
Comment (adithyaov) [inline Benchmark/Coreutils.hs:51]: do block indentation.

File: src/Streamly/Coreutils/Common.hs [hunk 31]
```diff
@@ -0,0 +1,49 @@
+module Streamly.Coreutils.Common (
+      someFileToFP
+    , someDirToFP
+   )
+where
+
+import Path
+import Path.Posix
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Common.hs:8]: Remove imports that are not used.

File: src/Streamly/Coreutils/Common.hs [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+-------------------------------------------------------------------------------
+-- helper functions for converting from SomeBase t to FilePath
+-------------------------------------------------------------------------------
+
+-- |
+-- Converts a type @SomeBase File@ to the type @FilePath@
+--
+-- @since 0.1.0.0
+someFileToFP :: SomeBase File -> FilePath
+someFileToFP some =
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Common.hs:35]: indentation

File: src/Streamly/Coreutils/Common.hs [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+                     Abs x -> fromAbsFile x
+                     Rel x -> fromRelFile x
+
+
+-- |
+-- Converts a type @SomeBase Dir@ to the type @FilePath@
+--
+-- @since 0.1.0.0
+someDirToFP :: SomeBase Dir -> FilePath
+someDirToFP some =
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Common.hs:46]: indentation

File: src/Streamly/Coreutils/Uniq.hs [hunk 34]
```diff
@@ -0,0 +1,217 @@
...
+            | All
+            -- ^
+            -- Allows all strings without filtering on the basis of counts
+
+
+-- |
+-- Data type of uniq's output
+--
+-- @since 0.1.0.0
+data UniqResult = Count Int String
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:44]: Rename `Count` to `UniqResult` instead.

File: src/Streamly/Coreutils/Uniq.hs [hunk 35]
```diff
@@ -0,0 +1,217 @@
...
+-- |
+-- Applies the comparison function to adjacent strings in the stream
+-- and returns a @Count Int String@ stream where the integer
+-- is the count of occurrences of the string
+--
+-- @since 0.1.0.0
+getRepetition :: (IsStream t, Monad m) => (String -> String -> Bool) -> t m String -> t m UniqResult
+getRepetition comparator = S.groupsBy comparator
+                              (FL.mkPureId (\(Count i a) s ->
+                                  if a == ""
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Uniq.hs:182]: Why do you need this if statement? Just `Count (i + 1) s` would be sufficient?

File: streamly-coreutils.cabal [hunk 36]
```diff
@@ -27,14 +27,97 @@ maintainer:          streamly@composewell.com
...
+  type: exitcode-stdio-1.0
+  hs-source-dirs: Benchmark/
+  main-is: Coreutils.hs
+  build-depends:
+      streamly-coreutils
+    , streamly
+    , base              >= 4.8   && < 5
+    , path              >= 0.8
+    , gauge             >= 0.2.4 && < 0.3
+    , monad-control     >= 1.0   && < 2
```
Comment (adithyaov) [inline streamly-coreutils.cabal:79]: `monad-control`?

File: src/Streamly/Coreutils/Common.hs [hunk 37]
```diff
@@ -0,0 +1,49 @@
+module Streamly.Coreutils.Common (
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Common.hs:1]: Remove this module if you are not using it for the changes in this PR.

File: streamly-coreutils.cabal [hunk 38]
```diff
@@ -27,14 +27,97 @@ maintainer:          streamly@composewell.com
...
+  type: exitcode-stdio-1.0
+  main-is: Coreutils.hs
+  js-sources: jsbits/clock.js
+  hs-source-dirs: test/Streamly
+  build-depends:
+      streamly-coreutils
+    , streamly
+    , base              >= 4.8   && < 5
+    , path              >= 0.8
+    , monad-control     >= 1.0   && < 2
```
Comment (adithyaov) [inline streamly-coreutils.cabal:102]: Where are you using `monad-control`?

File: streamly-coreutils.cabal [hunk 39]
```diff
@@ -27,14 +27,97 @@ maintainer:          streamly@composewell.com
...
+    , exceptions        >= 0.8   && < 0.11
+    , directory         >= 1.2.2 && < 1.4
+  default-language: Haskell2010
+
+-------------------------------------------------------------------------------
+-- Executables
+-------------------------------------------------------------------------------
+
+executable test-streaming
```
Comment (adithyaov) [inline streamly-coreutils.cabal:111]: What is this?

**Comment by adithyaov** [comment]
You can squash most of the commits. The only logical changes I see are:
1. Init
2. Uniq
3. Uniq tests
4. Uniq benchmarks
Try to keep the history clean. You can do this at the end though.

File: src/Streamly/Coreutils/Uniq.hs (hunk 35 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:182]: Actually, after comparison, if two strings are found the same, the previous and original one is taken. That's why.

File: streamly-coreutils.cabal (hunk 39 above)
Comment (wildcat26) [reply streamly-coreutils.cabal:111]: To check if the implementation is streaming, using infinite streams. Should I not push this ?

File: src/Streamly/Coreutils/Common.hs (hunk 37 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Common.hs:1]: Can I just re-export `Streamly.Coreutils.Uniq`?

Comment (wildcat26) [reply src/Streamly/Coreutils/Common.hs:1]: okay.

File: streamly-coreutils.cabal (hunk 38 above)
Comment (wildcat26) [reply streamly-coreutils.cabal:102]: Not using it. Removed.

File: streamly-coreutils.cabal (hunk 36 above)
Comment (wildcat26) [reply streamly-coreutils.cabal:79]: not needed, I was mistaken.

File: src/Streamly/Coreutils/Common.hs (hunk 31 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Common.hs:8]: Entire `Common.hs` unused in this PR

File: src/Streamly/Coreutils/Uniq.hs (hunk 34 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs:44]: :+1:

File: streamly-coreutils.cabal (hunk 39 above)
Comment (adithyaov) [reply streamly-coreutils.cabal:102]: No. We can write a test for that. Let it be for now.

File: src/Streamly/Coreutils/Uniq.hs (hunk 35 above)
Comment (adithyaov) [reply src/Streamly/Coreutils/Uniq.hs]: > the previous and original one is taken

What?

File: streamly-coreutils.cabal (hunk 39 above)
Comment (wildcat26) [reply streamly-coreutils.cabal:102]: :+1: okay letting it be

File: src/Streamly/Coreutils/Uniq.hs (hunk 35 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Uniq.hs]: Let's say with `skip fields = 1, two strings "field1 ye" and "field2 ye"` would be declared the same by the comparator. Then the stream should have `UniqResult 2 "field1 ye"` (first and original)

**Comment by adithyaov** [comment]
Ok, I think this is in good shape for the first iteration.
I'll make a few formatting changes and push it to this branch, then we can ask @harendra-kumar for a review and merge it.

File: streamly-coreutils.cabal [hunk 40]
```diff
@@ -27,14 +27,57 @@ maintainer:          streamly@composewell.com
...
+  import: compile-options
+  type: exitcode-stdio-1.0
+  hs-source-dirs: benchmark
+  main-is: Main.hs
+  build-depends:
+      streamly-coreutils
+    , streamly
+    , base              >= 4.8   && < 5
+    , gauge             >= 0.2.4 && < 0.3
+    , exceptions        >= 0.8   && < 0.11
```
Comment (adithyaov) [inline streamly-coreutils.cabal:63]: Should remove this.

**Comment by adithyaov** [comment]
Ok, I'm merging this PR. I'll rebase and force push to this branch.
